### PR TITLE
Build: Update Spark to 3.3.1

### PR DESCRIPTION
**Pull request by wangyum**
This PR update Spark to 3.3.1.

Spark 3.3.1 release notes:
https://spark.apache.org/releases/spark-release-3-3-1.html

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -96,6 +96,7 @@ allprojects {
   group = "org.apache.iceberg"
   version = projectVersion
   repositories {
+    maven { url  "https://repository.apache.org/content/repositories/orgapachespark-1418/" }
```
Comment (wangyum) [inline build.gradle:99]: Will remove it if Spark 3.3.1 is officially released.

Comment (wangyum) [reply build.gradle]: @Fokko Do you know how to run a single test? I'd like to run these tests locally:
```
org.apache.iceberg.spark.extensions.TestRewriteDataFilesProcedure > testRewriteDataFilesOnNonPartitionTable[catalogName = testhive, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hive, default-namespace=default}] FAILED
    java.lang.AssertionError: Action should rewrite 10 data files and add 1 data files: row 1 col 1 contents should match expected:<10> but was:<7>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.apache.iceberg.spark.SparkTestBase.assertEquals(SparkTestBase.java:181)
        at org.apache.iceberg.spark.SparkTestBase.assertEquals(SparkTestBase.java:163)
        at org.apache.iceberg.spark.extensions.TestRewriteDataFilesProcedure.testRewriteDataFilesOnNonPartitionTable(TestRewriteDataFilesProcedure.java:104)

org.apache.iceberg.spark.extensions.TestRewriteDataFilesProcedure > testRewriteDataFilesWithZOrder[catalogName = testhive, implementation = org.apache.iceberg.spark.SparkCatalog, config = {type=hive, default-namespace=default}] FAILED
    java.lang.AssertionError: Action should rewrite 10 data files and add 1 data files: row 1 col 1 contents should match expected:<10> but was:<7>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.apache.iceberg.spark.SparkTestBase.assertEquals(SparkTestBase.java:181)
        at org.apache.iceberg.spark.SparkTestBase.assertEquals(SparkTestBase.java:163)
        at org.apache.iceberg.spark.extensions.TestRewriteDataFilesProcedure.testRewriteDataFilesWithZOrder(TestRewriteDataFilesProcedure.java:171)
```

Comment (Fokko) [reply build.gradle]: @wangyum The easiest way is to use IntelliJ and just run a single test from the UI:
image

**Comment by bryanck** [comment]
I thought I'd mention that I uncovered a performance regression related to join distribution in Spark 3.3.0 vs Spark 3.2.x that impacts DSv2 sources like Iceberg. It would be nice if we could get a fix into Spark 3.3.1, but if not, we can work around it in the Iceberg Spark runtime. (The Spark ticket is [here](https://issues.apache.org/jira/browse/SPARK-40703) if interested.)

File: spark/v3.3/build.gradle [hunk 2]
```diff
@@ -27,6 +27,16 @@ def sparkProjects = [
...
 
+configure(sparkProjects) {
+  configurations {
+    all {
+      resolutionStrategy {
+        force 'com.fasterxml.jackson.core:jackson-databind:2.13.4.2'
+      }
+    }
+  }
+}
```
Comment (wangyum) [inline spark/v3.3/build.gradle:38]: To fix:
```
```
> Could not resolve all files for configuration ':iceberg-spark:iceberg-spark-3.3_2.12:compileClasspath'.

   > Could not find com.fasterxml.jackson:jackson-bom:2.13.4.1.
     Searched in the following locations:
       - https://repository.apache.org/content/repositories/orgapachespark-1430/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
       - https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
       - file:/home/runner/.m2/repository/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
     Required by:
         project :iceberg-spark:iceberg-spark-3.3_2.12 > org.apache.avro:avro:1.11.1 > com.fasterxml.jackson.core:jackson-databind:2.13.4.1
   > Could not find com.fasterxml.jackson:jackson-bom:2.13.4.1.
     Searched in the following locations:
       - https://repository.apache.org/content/repositories/orgapachespark-1430/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
       - https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
       - file:/home/runner/.m2/repository/com/fasterxml/jackson/jackson-bom/2.13.4.1/jackson-bom-2.13.4.1.pom
     Required by:
         project :iceberg-spark:iceberg-spark-3.3_2.12 > org.apache.arrow:arrow-vector:7.0.0 > com.fasterxml.jackson.core:jackson-annotations:2.13.4
         project :iceberg-spark:iceberg-spark-3.3_2.12 > org.apache.avro:avro:1.11.1 > com.fasterxml.jackson.core:jackson-core:2.13.4
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
```
```

**Comment by wangyum** [comment]
@bryanck The performance regression issue have been fixed.

**Review by singhpk234 — APPROVED** [review]
Thanks @wangyum !

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @wangyum! Thanks for reviewing, @singhpk234!

File: spark/v3.3/build.gradle (hunk 2 above)
Comment (flyrain) [reply spark/v3.3/build.gradle:38]: Thanks for the fix. The issue is that jackson-databind 2.13.4.1 depends on the non-existed lib, jackson-bom:2.13.4.1. I guess jackson team fixed that with jackson-databind 2.13.4.2

**Pull request merged**